File: clients/ios/BuildIt/Core/Nostr/CertificatePinning.swift
// CertificatePinning.swift
// BuildIt - Decentralized Mesh Communication
//
// Certificate pinning for Nostr relay connections to prevent MITM attacks.
// Supports pre-configured pins and Trust-on-First-Use (TOFU).

import Foundation
import CommonCrypto
import os.log

/// Certificate pinning configuration
struct CertPinConfig: Codable {
    /// Whether TOFU is enabled for unknown relays
    var tofuEnabled: Bool = true

    /// Whether to warn (vs block) when TOFU certificate changes
    var tofuWarnOnChange: Bool = true

    /// Whether write operations require pinned certificates
    var requirePinnedForWrite: Bool = true

    /// Days before pins should be refreshed
    var pinExpiryDays: Int = 365
}

/// Configuration for a single relay's certificate pins
struct RelayPinConfig: Codable {
    /// Primary certificate pins (SHA-256 fingerprints)
    var pins: [String]

    /// Backup pins for certificate rotation
    var backupPins: [String]

    /// When this pin was last verified
    var lastVerified: Date?

    /// Optional notes
    var notes: String?

    enum CodingKeys: String, CodingKey {
        case pins
        case backupPins = "backup_pins"
        case lastVerified = "last_verified"
        case notes
    }
}

/// Result of certificate verification
enum CertVerifyResult {
    /// Certificate matched a known pin
    case pinned

    /// Certificate matched a TOFU pin
    case tofu

    /// First use of this certificate (TOFU)
    case tofuFirstUse

    /// TOFU certificate changed (warning mode)
    case tofuChanged(previous: String, current: String)
}

/// Errors from certificate pinning
enum CertPinError: Error, LocalizedError {
    case pinMismatch(host: String, expected: [String], actual: String)
    case tofuCertificateChanged(host: String, previous: String, current: String)
    case noCertificate
    case invalidCertificate(String)
    case storageError(String)
    case configError(String)

    var errorDescription: String? {
        switch self {
        case .pinMismatch(let host, let expected, let actual):
            return "Certificate pin mismatch for \(host): expected \(expected.joined(separator: ", ")), got \(actual)"
        case .tofuCertificateChanged(let host, let previous, let current):
            return "Certificate changed for TOFU host \(host): was \(previous), now \(current)"
        case .noCertificate:
            return "No certificate provided by server"
        case .invalidCertificate(let msg):
            return "Invalid certificate: \(msg)"
        case .storageError(let msg):
            return "Pin storage error: \(msg)"
        case .configError(let msg):
            return "Configuration error: \(msg)"
        }
    }
}

/// Certificate pin store and verification
class CertificatePinStore {
    private let logger = Logger(subsystem: "com.buildit", category: "CertificatePinning")

    /// Configuration
    private(set) var config: CertPinConfig

    /// Pre-configured relay pins
    private var knownPins: [String: RelayPinConfig] = [:]

    /// TOFU pins (learned at runtime)
    private var tofuPins: [String: String] = [:]

    /// Queue for thread-safe access
    private let queue = DispatchQueue(label: "com.buildit.certpinstore")

    /// UserDefaults key for TOFU pins
    private let tofuDefaultsKey = "com.buildit.tofu.pins"

    /// Shared instance
    static let shared = CertificatePinStore()

    private init() {
        self.config = CertPinConfig()
        loadKnownPins()
        loadTofuPins()
    }

    /// Initialize with custom config
    init(config: CertPinConfig) {
        self.config = config
        loadKnownPins()
        loadTofuPins()
    }

    // MARK: - Pin Loading

    /// Load known pins from bundled configuration
    private func loadKnownPins() {
        // Load from bundled relay-pins.json if available
        if let url = Bundle.main.url(forResource: "relay-pins", withExtension: "json"),
           let data = try? Data(contentsOf: url),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let relays = json["relays"] as? [String: Any] {

            for (relayUrl, pinData) in relays {
                if let pinDict = pinData as? [String: Any] {
                    var pinConfig = RelayPinConfig(pins: [], backupPins: [])

                    if let pins = pinDict["pins"] as? [String] {
                        pinConfig.pins = pins
                    }
                    if let backupPins = pinDict["backup_pins"] as? [String] {
                        pinConfig.backupPins = backupPins
                    }
                    if let notes = pinDict["notes"] as? String {
                        pinConfig.notes = notes
                    }

                    knownPins[relayUrl] = pinConfig
                }
            }

            logger.info("Loaded \(self.knownPins.count) known relay pins")
        }
    }

    /// Load TOFU pins from UserDefaults
    private func loadTofuPins() {
        if let data = UserDefaults.standard.data(forKey: tofuDefaultsKey),
           let pins = try? JSONDecoder().decode([String: String].self, from: data) {
            tofuPins = pins
            logger.info("Loaded \(self.tofuPins.count) TOFU pins")
        }
    }

    /// Save TOFU pins to UserDefaults
    private func saveTofuPins() {
        if let data = try? JSONEncoder().encode(tofuPins) {
            UserDefaults.standard.set(data, forKey: tofuDefaultsKey)
        }
    }

    // MARK: - Pin Management

    /// Add a known relay pin
    func addKnownPin(url: String, config: RelayPinConfig) {
        queue.sync {
            knownPins[url] = config
        }
    }

    /// Compute SHA-256 fingerprint of a certificate
    static func computeFingerprint(_ certificateData: Data) -> String {
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        certificateData.withUnsafeBytes { bytes in
            _ = CC_SHA256(bytes.baseAddress, CC_LONG(certificateData.count), &hash)
        }
        return "sha256/" + Data(hash).base64EncodedString()
    }

    /// Compute fingerprint from SecCertificate
    static func computeFingerprint(_ certificate: SecCertificate) -> String? {
        guard let data = SecCertificateCopyData(certificate) as Data? else {
            return nil
        }
        return computeFingerprint(data)
    }

    // MARK: - Verification

    /// Verify a certificate against known or TOFU pins
    func verifyCertificate(host: String, certificateData: Data) -> Result<CertVerifyResult, CertPinError> {
        let fingerprint = Self.computeFingerprint(certificateData)
        let normalizedHost = normalizeHost(host)

        return queue.sync {
            // Check known pins first
            if let pinConfig = knownPins[normalizedHost] {
                let allPins = pinConfig.pins + pinConfig.backupPins

                if !allPins.isEmpty {
                    if allPins.contains(fingerprint) {
                        return .success(.pinned)
                    } else {
                        return .failure(.pinMismatch(
                            host: normalizedHost,
                            expected: allPins,
                            actual: fingerprint
                        ))
                    }
                }
                // Empty pins = known relay but no pins configured yet, fall through to TOFU
            }

            // Check TOFU pins
            if config.tofuEnabled {
                if let storedPin = tofuPins[normalizedHost] {
                    if storedPin == fingerprint {
                        return .success(.tofu)
                    } else {
                        // Certificate changed!
                        if config.tofuWarnOnChange {
                            logger.warning("Certificate changed for TOFU host \(normalizedHost): was \(storedPin), now \(fingerprint)")
                            return .success(.tofuChanged(previous: storedPin, current: fingerprint))
                        } else {
                            return .failure(.tofuCertificateChanged(
                                host: normalizedHost,
                                previous: storedPin,
                                current: fingerprint
                            ))
                        }
                    }
                }

                // First time seeing this host - store the pin
                tofuPins[normalizedHost] = fingerprint
                saveTofuPins()
                logger.info("TOFU: Stored initial certificate pin for \(normalizedHost): \(fingerprint)")
                return .success(.tofuFirstUse)
            }

            // No TOFU enabled and no known pins
            return .failure(.configError("No certificate pin configured for \(normalizedHost) and TOFU is disabled"))
        }
    }

    /// Verify a SecCertificate
    func verifyCertificate(host: String, certificate: SecCertificate) -> Result<CertVerifyResult, CertPinError> {
        guard let data = SecCertificateCopyData(certificate) as Data? else {
            return .failure(.invalidCertificate("Could not extract certificate data"))
        }
        return verifyCertificate(host: host, certificateData: data)
    }

    /// Check if a relay is pinned (known or TOFU)
    func isPinned(host: String) -> Bool {
        let normalizedHost = normalizeHost(host)

        return queue.sync {
            if let config = knownPins[normalizedHost], !config.pins.isEmpty {
                return true
            }
            return tofuPins[normalizedHost] != nil
        }
    }

    /// Clear TOFU pin for a specific host
    func clearTofuPin(host: String) {
        let normalizedHost = normalizeHost(host)
        queue.sync {
            tofuPins.removeValue(forKey: normalizedHost)
            saveTofuPins()
        }
    }

    /// Normalize host URL
    private func normalizeHost(_ host: String) -> String {
        if host.hasPrefix("wss://") || host.hasPrefix("ws://") {
            return host
        }
        return "wss://\(host)"
    }
}

// MARK: - URLSession Delegate for Certificate Pinning

/// URLSession delegate that implements certificate pinning
class PinnedURLSessionDelegate: NSObject, URLSessionDelegate {
    private let pinStore: CertificatePinStore
    private let logger = Logger(subsystem: "com.buildit", category: "CertificatePinning")

    /// Callback for certificate verification results
    var onCertificateVerified: ((String, CertVerifyResult) -> Void)?

    /// Callback for certificate verification failures
    var onCertificateError: ((String, CertPinError) -> Void)?

    init(pinStore: CertificatePinStore = .shared) {
        self.pinStore = pinStore
        super.init()
    }

    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let serverTrust = challenge.protectionSpace.serverTrust else {
            completionHandler(.performDefaultHandling, nil)
            return
        }

        let host = challenge.protectionSpace.host

        // Get the leaf certificate
        guard SecTrustGetCertificateCount(serverTrust) > 0 else {
            logger.error("No certificates in server trust for \(host)")
            onCertificateError?(host, .noCertificate)
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        // Get certificate at index 0 (leaf certificate)
        guard let certificate = SecTrustCopyCertificateChain(serverTrust)?[0] else {
            logger.error("Could not get leaf certificate for \(host)")
            onCertificateError?(host, .noCertificate)
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        // Verify against our pins
        let result = pinStore.verifyCertificate(host: host, certificate: certificate as! SecCertificate)

        switch result {
        case .success(let verifyResult):
            logger.info("Certificate verified for \(host): \(String(describing: verifyResult))")
            onCertificateVerified?(host, verifyResult)

            // Create credential and accept
            let credential = URLCredential(trust: serverTrust)
            completionHandler(.useCredential, credential)

        case .failure(let error):
            logger.error("Certificate pinning failed for \(host): \(error.localizedDescription)")
            onCertificateError?(host, error)
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }
}

// MARK: - WebSocket Task Delegate

/// WebSocket task delegate with certificate pinning
class PinnedWebSocketDelegate: NSObject, URLSessionWebSocketDelegate {
    private let pinStore: CertificatePinStore
    private let logger = Logger(subsystem: "com.buildit", category: "CertificatePinning")

    var onOpen: (() -> Void)?
    var onClose: ((URLSessionWebSocketTask.CloseCode, Data?) -> Void)?
    var onCertificateVerified: ((String, CertVerifyResult) -> Void)?
    var onCertificateError: ((String, CertPinError) -> Void)?

    init(pinStore: CertificatePinStore = .shared) {
        self.pinStore = pinStore
        super.init()
    }

    // Handle server trust challenge with certificate pinning
    nonisolated func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let serverTrust = challenge.protectionSpace.serverTrust else {
            completionHandler(.performDefaultHandling, nil)
            return
        }

        let host = challenge.protectionSpace.host

        // Get the certificate chain
        guard let certificateChain = SecTrustCopyCertificateChain(serverTrust),
              CFArrayGetCount(certificateChain) > 0 else {
            Task { @MainActor in
                self.logger.error("No certificates in server trust for \(host)")
                self.onCertificateError?(host, .noCertificate)
            }
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        // Get leaf certificate
        let certificate = (certificateChain as [AnyObject])[0] as! SecCertificate

        // Verify against our pins
        let result = pinStore.verifyCertificate(host: host, certificate: certificate)

        switch result {
        case .success(let verifyResult):
            Task { @MainActor in
                self.logger.info("Certificate verified for \(host)")
                self.onCertificateVerified?(host, verifyResult)
            }
            let credential = URLCredential(trust: serverTrust)
            completionHandler(.useCredential, credential)

        case .failure(let error):
            Task { @MainActor in
                self.logger.error("Certificate pinning failed for \(host): \(error.localizedDescription)")
                self.onCertificateError?(host, error)
            }
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }

    nonisolated func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol: String?
    ) {
        Task { @MainActor in
            self.onOpen?()
        }
    }

    nonisolated func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        Task { @MainActor in
            self.onClose?(closeCode, reason)
        }
    }
}

File: clients/android/app/src/main/java/network/buildit/core/nostr/CertificatePinning.kt
package network.buildit.core.nostr

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import android.util.Log
import okhttp3.CertificatePinner
import okhttp3.Handshake
import okhttp3.OkHttpClient
import org.json.JSONObject
import java.io.IOException
import java.security.MessageDigest
import java.security.cert.Certificate
import java.security.cert.X509Certificate
import javax.inject.Inject
import javax.inject.Singleton
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLPeerUnverifiedException
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

/**
 * Certificate pinning configuration.
 */
data class CertPinConfig(
    /** Whether TOFU is enabled for unknown relays */
    val tofuEnabled: Boolean = true,

    /** Whether to warn (vs block) when TOFU certificate changes */
    val tofuWarnOnChange: Boolean = true,

    /** Whether write operations require pinned certificates */
    val requirePinnedForWrite: Boolean = true,

    /** Days before pins should be refreshed */
    val pinExpiryDays: Int = 365
)

/**
 * Configuration for a single relay's certificate pins.
 */
data class RelayPinConfig(
    /** Primary certificate pins (SHA-256 fingerprints) */
    val pins: List<String>,

    /** Backup pins for certificate rotation */
    val backupPins: List<String>,

    /** When this pin was last verified */
    val lastVerified: Long? = null,

    /** Optional notes */
    val notes: String? = null
)

/**
 * Result of certificate verification.
 */
sealed class CertVerifyResult {
    /** Certificate matched a known pin */
    object Pinned : CertVerifyResult()

    /** Certificate matched a TOFU pin */
    object Tofu : CertVerifyResult()

    /** First use of this certificate (TOFU) */
    object TofuFirstUse : CertVerifyResult()

    /** TOFU certificate changed (warning mode) */
    data class TofuChanged(val previous: String, val current: String) : CertVerifyResult()
}

/**
 * Certificate pinning errors.
 */
sealed class CertPinError : Exception() {
    data class PinMismatch(
        val host: String,
        val expected: List<String>,
        val actual: String
    ) : CertPinError() {
        override val message: String
            get() = "Certificate pin mismatch for $host: expected ${expected.joinToString()}, got $actual"
    }

    data class TofuCertificateChanged(
        val host: String,
        val previous: String,
        val current: String
    ) : CertPinError() {
        override val message: String
            get() = "Certificate changed for TOFU host $host: was $previous, now $current"
    }

    object NoCertificate : CertPinError() {
        override val message: String
            get() = "No certificate provided by server"
    }

    data class InvalidCertificate(override val message: String) : CertPinError()

    data class StorageError(override val message: String) : CertPinError()

    data class ConfigError(override val message: String) : CertPinError()
}

/**
 * Certificate pin store and verification for Nostr relay connections.
 *
 * Implements certificate pinning to prevent MITM attacks:
 * - Pre-configured pins for known relays
 * - Trust-on-First-Use (TOFU) for unknown relays
 * - Backup pins for certificate rotation
 */
@Singleton
class CertificatePinStore @Inject constructor(
    private val context: Context
) {
    companion object {
        private const val TAG = "CertificatePinStore"
        private const val TOFU_PREFS_NAME = "buildit_tofu_pins"
        private const val TOFU_PINS_KEY = "pins"
    }

    /** Configuration */
    var config: CertPinConfig = CertPinConfig()
        private set

    /** Pre-configured relay pins */
    private val knownPins = mutableMapOf<String, RelayPinConfig>()

    /** TOFU pins (learned at runtime) */
    private val tofuPins = mutableMapOf<String, String>()

    /** SharedPreferences for TOFU persistence */
    private val tofuPrefs: SharedPreferences by lazy {
        context.getSharedPreferences(TOFU_PREFS_NAME, Context.MODE_PRIVATE)
    }

    init {
        loadKnownPins()
        loadTofuPins()
    }

    /**
     * Load known pins from bundled configuration.
     */
    private fun loadKnownPins() {
        try {
            val inputStream = context.assets.open("relay-pins.json")
            val jsonString = inputStream.bufferedReader().use { it.readText() }
            val json = JSONObject(jsonString)

            val relays = json.optJSONObject("relays") ?: return

            for (url in relays.keys()) {
                val pinData = relays.getJSONObject(url)
                val pins = mutableListOf<String>()
                val backupPins = mutableListOf<String>()

                pinData.optJSONArray("pins")?.let { arr ->
                    for (i in 0 until arr.length()) {
                        pins.add(arr.getString(i))
                    }
                }

                pinData.optJSONArray("backup_pins")?.let { arr ->
                    for (i in 0 until arr.length()) {
                        backupPins.add(arr.getString(i))
                    }
                }

                knownPins[url] = RelayPinConfig(
                    pins = pins,
                    backupPins = backupPins,
                    notes = pinData.optString("notes", null)
                )
            }

            Log.i(TAG, "Loaded ${knownPins.size} known relay pins")
        } catch (e: IOException) {
            Log.w(TAG, "Could not load relay-pins.json: ${e.message}")
        } catch (e: Exception) {
            Log.e(TAG, "Error loading known pins", e)
        }
    }

    /**
     * Load TOFU pins from SharedPreferences.
     */
    private fun loadTofuPins() {
        val pinsJson = tofuPrefs.getString(TOFU_PINS_KEY, null) ?: return
        try {
            val json = JSONObject(pinsJson)
            for (key in json.keys()) {
                tofuPins[key] = json.getString(key)
            }
            Log.i(TAG, "Loaded ${tofuPins.size} TOFU pins")
        } catch (e: Exception) {
            Log.e(TAG, "Error loading TOFU pins", e)
        }
    }

    /**
     * Save TOFU pins to SharedPreferences.
     */
    private fun saveTofuPins() {
        val json = JSONObject()
        tofuPins.forEach { (key, value) -> json.put(key, value) }
        tofuPrefs.edit().putString(TOFU_PINS_KEY, json.toString()).apply()
    }

    /**
     * Add a known relay pin.
     */
    fun addKnownPin(url: String, config: RelayPinConfig) {
        synchronized(knownPins) {
            knownPins[url] = config
        }
    }

    /**
     * Compute SHA-256 fingerprint of a certificate.
     */
    fun computeFingerprint(certificate: Certificate): String {
        val encoded = certificate.encoded
        val digest = MessageDigest.getInstance("SHA-256").digest(encoded)
        return "sha256/" + Base64.encodeToString(digest, Base64.NO_WRAP)
    }

    /**
     * Compute SHA-256 fingerprint of raw certificate data.
     */
    fun computeFingerprint(certificateData: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(certificateData)
        return "sha256/" + Base64.encodeToString(digest, Base64.NO_WRAP)
    }

    /**
     * Verify a certificate against known or TOFU pins.
     */
    fun verifyCertificate(host: String, certificate: Certificate): Result<CertVerifyResult> {
        val fingerprint = computeFingerprint(certificate)
        val normalizedHost = normalizeHost(host)

        return synchronized(this) {
            // Check known pins first
            knownPins[normalizedHost]?.let { pinConfig ->
                val allPins = pinConfig.pins + pinConfig.backupPins

                if (allPins.isNotEmpty()) {
                    if (fingerprint in allPins) {
                        return@synchronized Result.success(CertVerifyResult.Pinned)
                    } else {
                        return@synchronized Result.failure(
                            CertPinError.PinMismatch(
                                host = normalizedHost,
                                expected = allPins,
                                actual = fingerprint
                            )
                        )
                    }
                }
                // Empty pins = known relay but no pins configured yet, fall through to TOFU
            }

            // Check TOFU pins
            if (config.tofuEnabled) {
                tofuPins[normalizedHost]?.let { storedPin ->
                    if (storedPin == fingerprint) {
                        return@synchronized Result.success(CertVerifyResult.Tofu)
                    } else {
                        // Certificate changed!
                        if (config.tofuWarnOnChange) {
                            Log.w(
                                TAG,
                                "Certificate changed for TOFU host $normalizedHost: was $storedPin, now $fingerprint"
                            )
                            return@synchronized Result.success(
                                CertVerifyResult.TofuChanged(
                                    previous = storedPin,
                                    current = fingerprint
                                )
                            )
                        } else {
                            return@synchronized Result.failure(
                                CertPinError.TofuCertificateChanged(
                                    host = normalizedHost,
                                    previous = storedPin,
                                    current = fingerprint
                                )
                            )
                        }
                    }
                }

                // First time seeing this host - store the pin
                tofuPins[normalizedHost] = fingerprint
                saveTofuPins()
                Log.i(TAG, "TOFU: Stored initial certificate pin for $normalizedHost: $fingerprint")
                return@synchronized Result.success(CertVerifyResult.TofuFirstUse)
            }

            // No TOFU enabled and no known pins
            Result.failure(
                CertPinError.ConfigError(
                    "No certificate pin configured for $normalizedHost and TOFU is disabled"
                )
            )
        }
    }

    /**
     * Check if a relay is pinned (known or TOFU).
     */
    fun isPinned(host: String): Boolean {
        val normalizedHost = normalizeHost(host)
        return synchronized(this) {
            knownPins[normalizedHost]?.pins?.isNotEmpty() == true ||
                    tofuPins.containsKey(normalizedHost)
        }
    }

    /**
     * Clear TOFU pin for a specific host.
     */
    fun clearTofuPin(host: String) {
        val normalizedHost = normalizeHost(host)
        synchronized(this) {
            tofuPins.remove(normalizedHost)
            saveTofuPins()
        }
    }

    /**
     * Normalize host URL.
     */
    private fun normalizeHost(host: String): String {
        return if (host.startsWith("wss://") || host.startsWith("ws://")) {
            host
        } else {
            "wss://$host"
        }
    }

    /**
     * Build an OkHttpClient with certificate pinning for known relays.
     *
     * Note: This only pins known relays. TOFU verification happens in
     * the custom TrustManager.
     */
    fun buildPinnedOkHttpClient(baseBuilder: OkHttpClient.Builder = OkHttpClient.Builder()): OkHttpClient {
        // Build CertificatePinner for known relays with configured pins
        val pinnerBuilder = CertificatePinner.Builder()

        synchronized(knownPins) {
            for ((url, config) in knownPins) {
                val allPins = config.pins + config.backupPins
                if (allPins.isNotEmpty()) {
                    // Extract hostname from URL
                    val hostname = url
                        .removePrefix("wss://")
                        .removePrefix("ws://")
                        .removeSuffix("/")
                        .split("/")[0]

                    for (pin in allPins) {
                        pinnerBuilder.add(hostname, pin)
                    }
                }
            }
        }

        // Create custom TrustManager for TOFU
        val tofuTrustManager = TofuTrustManager(this)

        val sslContext = SSLContext.getInstance("TLS")
        sslContext.init(null, arrayOf<TrustManager>(tofuTrustManager), null)

        return baseBuilder
            .certificatePinner(pinnerBuilder.build())
            .sslSocketFactory(sslContext.socketFactory, tofuTrustManager)
            .hostnameVerifier(TofuHostnameVerifier(this))
            .build()
    }
}

/**
 * Custom X509TrustManager that implements TOFU.
 */
class TofuTrustManager(
    private val pinStore: CertificatePinStore
) : X509TrustManager {

    companion object {
        private const val TAG = "TofuTrustManager"
    }

    // Get the default trust manager
    private val defaultTrustManager: X509TrustManager by lazy {
        val trustManagerFactory =
            javax.net.ssl.TrustManagerFactory.getInstance(javax.net.ssl.TrustManagerFactory.getDefaultAlgorithm())
        trustManagerFactory.init(null as java.security.KeyStore?)
        trustManagerFactory.trustManagers
            .filterIsInstance<X509TrustManager>()
            .first()
    }

    override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {
        defaultTrustManager.checkClientTrusted(chain, authType)
    }

    override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {
        // First, do default certificate chain validation
        defaultTrustManager.checkServerTrusted(chain, authType)

        // Then verify against our TOFU pins
        // Note: Known pins are handled by CertificatePinner in OkHttp
        if (chain.isNullOrEmpty()) {
            throw SSLPeerUnverifiedException("No certificates provided")
        }

        val leafCert = chain[0]

        // We'll verify in the hostname verifier where we have the host
        // Store the certificate temporarily for verification
        Log.d(TAG, "Server certificate validated by chain, TOFU check deferred to hostname verifier")
    }

    override fun getAcceptedIssuers(): Array<X509Certificate> {
        return defaultTrustManager.acceptedIssuers
    }
}

/**
 * Hostname verifier that implements TOFU verification.
 */
class TofuHostnameVerifier(
    private val pinStore: CertificatePinStore
) : HostnameVerifier {

    companion object {
        private const val TAG = "TofuHostnameVerifier"
    }

    // Default hostname verifier
    private val defaultVerifier = javax.net.ssl.HttpsURLConnection.getDefaultHostnameVerifier()

    override fun verify(hostname: String, session: javax.net.ssl.SSLSession): Boolean {
        // First, do default hostname verification
        if (!defaultVerifier.verify(hostname, session)) {
            Log.e(TAG, "Default hostname verification failed for $hostname")
            return false
        }

        // Get the certificate
        val certificates = try {
            session.peerCertificates
        } catch (e: SSLPeerUnverifiedException) {
            Log.e(TAG, "No peer certificates for $hostname")
            return false
        }

        if (certificates.isEmpty()) {
            Log.e(TAG, "Empty certificate chain for $hostname")
            return false
        }

        val leafCert = certificates[0]

        // Verify against TOFU pins
        val result = pinStore.verifyCertificate(hostname, leafCert)

        return result.fold(
            onSuccess = { verifyResult ->
                when (verifyResult) {
                    is CertVerifyResult.Pinned -> {
                        Log.i(TAG, "Certificate verified (pinned) for $hostname")
                        true
                    }
                    is CertVerifyResult.Tofu -> {
                        Log.i(TAG, "Certificate verified (TOFU) for $hostname")
                        true
                    }
                    is CertVerifyResult.TofuFirstUse -> {
                        Log.i(TAG, "First use certificate stored (TOFU) for $hostname")
                        true
                    }
                    is CertVerifyResult.TofuChanged -> {
                        Log.w(
                            TAG,
                            "Certificate changed for $hostname (TOFU warning): " +
                                    "${verifyResult.previous} -> ${verifyResult.current}"
                        )
                        // Allow but warn - caller should handle this
                        true
                    }
                }
            },
            onFailure = { error ->
                Log.e(TAG, "Certificate verification failed for $hostname: ${error.message}")
                false
            }
        )
    }
}

/**
 * Callback interface for certificate pinning events.
 */
interface CertificatePinningCallback {
    /** Called when a certificate is verified successfully */
    fun onCertificateVerified(host: String, result: CertVerifyResult)

    /** Called when certificate verification fails */
    fun onCertificateError(host: String, error: CertPinError)
}
